### PR TITLE
SALTO-1018 Convert profile list fields to maps

### PR DESCRIPTION
**Pull request by netama**
Convert profile list fields into maps - initially with an option to switch back to the old format (`useOldProfiles`).

The initial configuration supports a map with up to two levels, based on the value of a specific field (converted to `naclCase`).
Includes support for maps-of-lists, for cases where multiple items are mapped to the same map key. This will ideally be configured up front (like we do for `layoutAssignments`, which can have multiple items under the same layout with different `recordType`s), but is also a fallback for _all_ fields, in order to avoid failing the fetch in case one of the uniqueness assumptions fails.

https://github.com/salto-io/salto/pull/1561 has a change validator that verifies that the map keys were not modified manually (and provides the correct keys to use in case of inconsistencies).

---
_Release Notes_

Updated profile structure for easier navigation and lookup - replacing list fields with maps.

_Upgrade Instructions_

Run fetch before deploy.

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+  Object.values(profileObj.fields).filter(
+    f => profileMapFielDef[f.name] !== undefined && isMapType(f.type)
+  ).forEach(f => {
+    if (isMapType(f.type)) {
+      f.type = f.type.innerType
+    }
+    // for nested fields (not using while to avoid edge cases)
+    if (isMapType(f.type)) {
+      f.type = f.type.innerType
+    }
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:282]: isn't this missing the part where we create a list type?
if there are no duplicates the `innerType` of the map would be an object type, no?

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    convertFieldsBackToLists(profileInstanceChanges, PROFILE_MAP_FIELD_DEF)
+    convertFieldTypesBackToLists(
+      profileObj,
+      PROFILE_MAP_FIELD_DEF,
+    )
+  },
+
+  onDeploy: async changes => {
+    if (config.useOldProfiles) {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:339]: preDeploy doesn't check this, so it will convert values from maps to lists.
if we don't convert back here it will create a lot of unexpected changes to the instance that was deployed and cause a lot of validation errors.

if we want to support a mismatch between the configuration of `useOldProfiles` and the values, it should be supported here as well, preDeploy and onDeploy should by symmetrical

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    const profileInstanceChanges = getInstanceChanges(changes)
+    if (profileInstanceChanges.length === 0) {
+      return []
+    }
+    const profileObj = getChangeElement(profileInstanceChanges[0]).type
+    // after preDeploy, the fields with lists are exactly the ones that should be converted
+    // back to lists
+    const nonUniqueMapFields = Object.keys(profileObj.fields).filter(
+      fieldName => isListType((profileObj.fields[fieldName].type))
+    )
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:352]: does this mean we are actually relying on `convertFieldTypesBackToLists` generating the wrote field types (not creating list fields for fields that have list values)?
if we are intentionally leaving the type inaccurate, why do we need to change it at all? wouldn't it be simpler to just keep it as is in preDeploy and here?

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    const modifiedProfileInstances = profileInstanceChanges.map(getChangeElement)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:353]: in a `modify` change this will only fix the `after` instance and it will leave the `before` instance with lists (because `preDeploy` does handle both

File: packages/salesforce-adapter/src/filters/profile_maps.ts (hunk 3 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/profile_maps.ts]: I tried to leave the type unmodified (with the map definitions) - that fails because we use `transformElement` in [toDeployableInstance](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/src/transformers/transformer.ts#L1173) (and I think even if it did work somehow, it could cause us not to omit `localOnly` fields, so we shouldn't do it). I'm leaving the commit with the unmodified type, you can see the e2e failure for it too if that's interesting. but I don't think we can work around modifying the type.

the other (ugly) option is to add synthetic hidden fields that will be used as placeholders for the original values during the `preDeploy-onDeploy` period - I'm assuming that's not better?

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+  changes
+    .filter(isAdditionOrModificationChange)
+    .filter(isInstanceChange)
+    .filter(change => metadataType(getChangeElement(change)) === PROFILE_METADATA_TYPE)
+)
+
+const findProfileInstances = (
+  elements: Iterable<Element>,
+): InstanceElement[] => {
+  const instances = wu(elements).filter(isInstanceElement) as wu.WuIterable<InstanceElement>
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:281]: we should avoid casts when possible, in this case I don't see why the cast is needed.
(also, not sure why the random switch to iterables here)

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+  onDeploy: async changes => {
+    if (config.useOldProfiles) {
+      return []
+    }
+
+    const profileInstanceChanges = getInstanceChanges(changes)
+    if (profileInstanceChanges.length === 0) {
+      return []
+    }
+    // since salesforce does not require list fields to be defined as lists,
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:329]: IIUC this works because in `transformElement` we ignore list value / type mismatch (it was supposed to be a temporary "hack" when we introduced list types to avoid having to fix all the types in that one PR, but seems like it can't really be undone anymore...), not because of salesforce.
also, sounds like this comment should be in `preDeploy`

File: packages/salesforce-adapter/src/filters/profile_maps.ts [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    // we only mark fields as lists of their map inner value is a list,
+    // so that we can convert the object back correctly in onDeploy
+    convertInstanceFieldsToMaps(
+      profileInstanceChanges.flatMap(getAllChangeElements),
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/profile_maps.ts:333]: I don't like this new method of flattening `getAllChangeElements`, it feels more side-effecty than `applyFunctionToChangeData` because at least the latter is at least explicit about applying a change in-place, the former just "gets", making the side effect that will happen later much more hidden and hard to follow IMO.

it is also asymmetrical with `preDeploy` which does use `applyFunctionToChangeData`

**Review by ori-moisis — APPROVED** [review]

File: packages/salesforce-adapter/src/filters/profile_maps.ts (hunk 5 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/profile_maps.ts:281]: it was taken from `findInstances` because it ended up not fitting here (was using `e.type.elemID` instead of `metadataType`). the cast is needed, but I'll get rid of `wu`.

File: packages/salesforce-adapter/src/filters/profile_maps.ts (hunk 6 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/profile_maps.ts:329]: I actually meant the salesforce side, but you're right that it's both. I'll update the comment (and move it)
